language/data-types/pcore-data-representation.md

```diff
@@ -0,0 +1,58 @@
+Pcore Data Representation
+===
+Puppet's type system (Pcore) has two kinds of data representation (in addition to being able to describe data in the Puppet Programming Language itself).
+
+* [Pcore Generic Data][1] - As a programming language specific but generic data structure (that can then be
+  serialized using a technology of choice for that programming language).
+* [Pcore Serialization][2] - As a Pcore specific encoding using either a textual or binary "on the wire" format.
+
+Important Note about Circular Data
+---
+Regular data in Puppet is never circular since to all values are immutable and thus making it
+impossible to create such a structure. This means that care must be taken in custom Ruby logic
+that is integrated with Puppet to ensure that such custom logic does not return circular data.
+
+While regular data in Puppet cannot be circular, definitions of data types can be. Consider
+the type `type StringTree = Array[Variant[String, StringTree]]` - when this type is serialized
+(as opposed to a serialization of a value of this type) it clearly needs to be able to
+refer back to itself recursively.
+
+There are features in Pcore data representation to handle circular definitions. Either:
+
+* Avoid the issue all together by:
+  * using "type by reference = true" (a feature that does not include data types in the result)
+  * using data that is not circular (which it typically never is)
```

Comment:
I wonder if it wouldn't be easier to understand "using types that are not circular", since a) only types can be circular and b) the other items in this list talks about "type by reference".